krystex/src/main/java/com/flipkart/krystal/krystex/node/KrystalNodeExecutor.java

```diff
@@ -179,19 +186,47 @@ private void createDependantNodes(NodeId nodeId, DependantChain dependantChain)
     }
   }
 
-  CompletableFuture<NodeResponse> enqueueCommand(NodeRequestCommand nodeCommand) {
-    return supplyAsync(
-            () -> nodeRegistry.get(nodeCommand.nodeId()).executeRequestCommand(nodeCommand),
-            commandQueueLease.get())
-        .thenCompose(identity());
+  /**
+   * Enqueues the provided NodeRequestCommand supplier into the command queue. This method is
+   * intended to be called in threads other than the main thread of this KrystalNodeExecutor.(for
+   * example IO reactor threads). When a non-blocking IO call is made by a Node, a callback is added
+   * to the resulting CompletableFuture which generates an ExecuteWithDependency command for its
+   * dependents. That is when this method is used - ensuring that all further processing of the
+   * nodeCammand happens in the main thread.
+   */
+  CompletableFuture<NodeResponse> enqueueNodeCommand(Supplier<NodeRequestCommand> nodeCommand) {
+    return enqueueCommand(() -> _executeCommand(nodeCommand.get())).thenCompose(identity());
+  }
+
+  /**
+   * This method can be called only from the main thread of this KrystalNodeExecutor. Calling this
+   * method from any other thread (for example: IO reactor threads) will cause race conditions,
+   * multithreaded access of non-thread-safe data structures, and resulting unspecified behaviour.
+   *
+   * <p>This is an optimal version on {@link #enqueueNodeCommand(Supplier<NodeRequestCommand>)}
+   * which bypasses the command queue for the special case that the command is originating from the
+   * same main thread inside the command queue,thus avoiding unnecessary contention in the
+   * thread-safe structures inside the command queue.
+   */
+  CompletableFuture<NodeResponse> executeCommand(NodeCommand nodeCommand) {
+    krystalNodeMetrics.commandQueueBypassed();
+    return _executeCommand(nodeCommand);
   }
 
-  void enqueueCommand(Flush flush) {
-    runAsync(() -> nodeRegistry.get(flush.nodeId()).executeCommand(flush), commandQueueLease.get());
+  private CompletableFuture<NodeResponse> _executeCommand(NodeCommand nodeCommand) {
+    if (nodeCommand instanceof NodeRequestCommand nodeRequestCommand) {
+      return nodeRegistry.get(nodeCommand.nodeId()).executeRequestCommand(nodeRequestCommand);
+    } else if (nodeCommand instanceof Flush flush) {
+      nodeRegistry.get(flush.nodeId()).executeCommand(flush);
+      return failedFuture(new UnsupportedOperationException("No data returned for flush command"));
```

Comment:
This block always returns failedFuture for Flush command. Is it correct?

Reply:
Yes, this is intentional. This is done to consolidate all command executions in a single method for ease of metric collection and code understanding. Unfortunately flush is a special command which doesnot generate a NodeResponse. Hence returning failedFuture - anyway callers sending a Flush command will not access the future. So this should be ok. If any caller does access it, it will fail.

If this becomes problematic in future, we can create a `empty` NodeResponse record and return it. But not going to that lengths because this is anyway a platform internal method.